sympy/polys/rings.py

```diff
@@ -3005,3 +3032,222 @@ def sqf_list(f, all=False):
 
     def factor_list(f):
         return f.ring.dmp_factor_list(f)
+
+def monomial_extract(polynomials):
+    """
+    Extracts any common monomial from the polynomials in p.
+
+    Examples
+    ========
+
+    >>> from sympy.polys.rings import monomial_extract
+    >>> from sympy import ZZ, ring
+    >>> R, x, y = ring("x, y", ZZ)
+
+    >>> f = x + y
+    >>> g = x**2*y + x*y
+    >>> polynomials = [f, g]
+    >>> monomial_extract(polynomials)
+    ([x + y, x**2*y + x*y], None)
+
+    >>> f = x**2*y
+    >>> g = x**2*y + x*y
+    >>> polynomials = [f, g]
+    >>> monomial_extract(polynomials)
+    ([x, x + 1], x*y)
+
+    """
+    ring = polynomials[0].ring
+    zero_monoms = ring.zero_monom
+    monoms = chain(*polynomials)
+    monom_gcd = tuple(map(min, zip(*monoms)))
+
+    # If the zero monomial is present, we can immediately return
+    if (any(zero_monoms in poly for poly in polynomials)
```

Comment:
The check for `any(zero_monom in poly ...)` should be done *before* computing `monom_gcd` by looping over all monomials of all polynomials. If any polynomial has the `zero_monom` then there is no need to loop over any of the other monomials.

Also don't use `zero_monoms` (plural) when it should be `zero_monom` singular.